proxy/src/control/cache.rs

```diff
@@ -37,54 +46,85 @@ impl<T> Exists<T> {
             reset_on_next_modification: true,
         }
     }
 
-    pub fn values(&self) -> &HashSet<T> {
-        &self.values
-    }
-
     pub fn set_reset_on_next_modification(&mut self) {
         self.reset_on_next_modification = true;
     }
 
-    pub fn extend<I, F>(&mut self, iter: I, on_change: &mut F)
-        where I: Iterator<Item = T>,
-              F: FnMut(T, CacheChange),
+    /// Update the cache to contain the union of its current contents and the
+    /// key-value pairs in `iter`. Pairs not present in the cache will be
+    /// inserted, and keys present in both the cache and the iterator will be
+    /// updated so that their values match those in the iterator.
+    pub fn update_union<I, F>(&mut self, iter: I, on_change: &mut F)
+    where
+        I: Iterator<Item = (K, V)>,
+        F: FnMut((K, V), CacheChange),
     {
-        fn extend_inner<T, I, F>(values: &mut HashSet<T>, iter: I, on_change: &mut F)
-            where T: Copy + Eq + std::hash::Hash, I: Iterator<Item = T>, F: FnMut(T, CacheChange)
+        fn update_inner<K, V, I, F>(
+            values: &mut IndexMap<K, V>,
+            iter: I,
+            on_change: &mut F
+        )
+        where
+            K: Eq + Hash + Copy + Clone,
+            V: PartialEq + Clone,
+            I: Iterator<Item = (K, V)>,
+            F: FnMut((K, V), CacheChange),
         {
-            for value in iter {
-                if values.insert(value) {
-                    on_change(value, CacheChange::Insertion);
+            for (key, value) in iter {
+                match values.insert(key, value.clone()) {
+                    // If the returned value is equal to the inserted value,
+                    // then the inserted key was already present in the map
+                    // and the value is unchanged. Do nothing.
+                    Some(ref old_value) if old_value == &value => {},
+                    // If `insert` returns `Some` with a different value than
+                    // the one we inserted, then we changed the value for that
+                    // key.
+                    Some(_) =>
+                        on_change((key, value), CacheChange::Modification),
+                    // If `insert` returns `None`, then there was no old value
+                    // previously present. Therefore, we inserted a new value
+                    // into the cache.
+                     None => on_change((key, value), CacheChange::Insertion),
                 }
             }
         }
 
         if !self.reset_on_next_modification {
-            extend_inner(&mut self.values, iter, on_change);
+            update_inner(&mut self.values, iter, on_change);
         } else {
-            let to_insert = iter.collect::<HashSet<T>>();
-            extend_inner(&mut self.values, to_insert.iter().map(|value| *value), on_change);
-            self.retain(&to_insert, on_change);
+            let to_insert = iter.collect::<IndexMap<K, V>>();
```

Comment:
Does it make sense to use `IndexMap` here? `Vec<(K, V)>` seems sufficient and probably slightly more efficient. I think I used `HashSet` previously because `retain()` required a HashSet.

Reply:
@briansmith this is an `IndexMap` so that it can be passed to `update_intersection`, which takes an `&mut IndexMap`